runtime/src/main/java/org/corfudb/runtime/view/TableRegistry.java

```diff
@@ -878,4 +887,198 @@ public void shutdown() {
+    public static class FullyQualifiedTableName {
+        @Default
+        private final Optional<String> namespace = Optional.empty();
+        private final String tableName;
+
+        /**
+         * Fully qualified table name created to produce the stream uuid.
+         *
+         * @return Fully qualified table name.
+         */
+        public String toFqdn() {
+            return namespace
+                    .map(ns -> ns + "$" + tableName)
+                    .orElse(tableName);
+        }
+
+        public StreamId toStreamId() {
+            return StreamId.build(toFqdn());
+        }
+
+        public StreamName toStreamName() {
+            return StreamName.build(toFqdn());
+        }
+
+        public static StreamId streamId(String ns, String name) {
+            return build(ns, name).toStreamId();
+        }
+
+        public static FullyQualifiedTableName build(String ns, String name) {
+            Optional<String> maybeNs;
+            if (StringUtils.isEmpty(ns)) {
+                maybeNs = Optional.empty();
+            } else {
+                maybeNs = Optional.of(ns);
+            }
+
+            return FullyQualifiedTableName.builder()
+                    .namespace(maybeNs)
+                    .tableName(name)
+                    .build();
+        }
+
+        public static StreamId streamId(TableName tableName) {
+            return build(tableName).toStreamId();
+        }
+
+        /**
+         * Fully qualified table name created to produce the stream uuid.
+         *
+         * @param tableName TableName of the table.
+         * @return Fully qualified table name.
+         */
+        public static FullyQualifiedTableName build(TableName tableName) {
+            Optional<String> ns = Optional
+                    .of(tableName.getNamespace())
+                    .filter(Predicate.not(String::isEmpty));
+
+            return FullyQualifiedTableName.builder()
+                    .tableName(tableName.getTableName())
+                    .namespace(ns)
+                    .build();
+        }
+
+        public TableName toTableName() {
+            return TableName.newBuilder()
+                    .setNamespace(rawNamespace())
+                    .setTableName(tableName)
+                    .build();
+        }
+
+        public boolean isSystemNs(){
+            return namespace
+                    .map(ns -> ns.equals(CORFU_SYSTEM_NAMESPACE))
+                    .orElse(false);
+        }
+
+        public boolean isSystemTable() {
+            boolean isDescriptorTable = tableName.equals(TableRegistry.PROTOBUF_DESCRIPTOR_TABLE_NAME);
+            boolean isRegistryTable = tableName.equals(TableRegistry.REGISTRY_TABLE_NAME);
+            return isDescriptorTable || isRegistryTable;
+        }
+
+        public boolean isSystemNsAndTable() {
+            return isSystemNs() && isSystemTable();
+        }
+
+        public String rawNamespace() {
+            return namespace.orElse("");
```

Comment:
if null is being used for some corfu table's namespace there is likely some bug - can you please share what prompted you to use an optional here?

Reply:
The optional is exactly to prevent us from having NPE and enforce code to always consider an empty namespace. Unfortunately I have to be consistent also with the old approach, so I have to return an empty string here (as far as i remember it's only for tests).

And we check for null and empty string https://github.com/CorfuDB/CorfuDB/blob/123a2445fb80b6abda1aa3d36249bc5661c74fd2/runtime/src/main/java/org/corfudb/runtime/view/TableRegistry.java#L1019

plus we check it in builder.

Also, I forgot to mention in my prev comments, now bacuse we have those classes and builders we really check all the parameters upfront, in compare to `@Nonnull` annotations in methods parameters which as you can remebers simply don't work and bring a big risk

Reply:
I understand the `@Nonnull` issue but I am really curious to know why is there a need for null in the namespace anyway - I am afraid it could be masking some other existing bug in the system that we should address.

Reply:
So, I guess my answer is still valid here also https://github.com/CorfuDB/CorfuDB/pull/3977#discussion_r1633577137